src/main/java/org/openrewrite/java/migrate/javax/AddScopeToInjectedClass.java

```diff
@@ -0,0 +1,96 @@
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+public class AddScopeToInjectedClass extends ScanningRecipe<Set<JavaType.FullyQualified>> {
+    private static final String JAVAX_INJECT_INJECT = "javax.inject.Inject";
+    private static final String JAVAX_ENTERPRISE_CONTEXT_DEPENDENT = "javax.enterprise.context.Dependent";
+    private static final Collection<String> TYPES_PROMPTING_SCOPE_ADDITION = Arrays.asList(JAVAX_INJECT_INJECT);
+
+    @Override
+    public String getDisplayName() {
+        return "Add scope annotation to injected classes";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Finds member variables annotated with `@Inject' and applies `@Dependent` scope annotation to the variable's type.";
+    }
+
+    @Override
+    public Set<JavaType.FullyQualified> getInitialValue(ExecutionContext ctx) {
+        return new HashSet<>();
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getScanner(Set<JavaType.FullyQualified> injectedTypes) {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, ExecutionContext executionContext) {
+                J.ClassDeclaration cd = super.visitClassDeclaration(classDecl, executionContext);
+                for (JavaType.Variable variable : cd.getType().getMembers()) {
+                    if (variableTypeRequiresScope(variable)) {
+                        injectedTypes.add((JavaType.FullyQualified) variable.getType());
+                    }
+                }
+                return cd;
+            }
+
+            private boolean variableTypeRequiresScope(@Nullable JavaType.Variable memberVariable) {
+                if (memberVariable == null) {
+                    return false;
+                }
+
+                for (JavaType.FullyQualified annotation : memberVariable.getAnnotations()) {
+                    if (TYPES_PROMPTING_SCOPE_ADDITION.contains(annotation.getFullyQualifiedName())) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+        };
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor(Set<JavaType.FullyQualified> injectedTypes) {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.CompilationUnit visitCompilationUnit(J.CompilationUnit compilationUnit, ExecutionContext executionContext) {
+                J.CompilationUnit cu = super.visitCompilationUnit(compilationUnit, executionContext);
+                for (J.ClassDeclaration aClass : cu.getClasses()) {
+                    if (injectedTypes.contains(aClass.getType())) {
```

Comment:
Please note that `JavaType.Class` does not implement `hashCode()` so you cannot really add these to a `HashSet`. This could be problematic when the recipe scheduler executes a recipe against a large repository, because then it breaks up the sources into "chunks", which each have their own set of `JavaType` objects (i.e. not referential equality).

For the same reason it isn't really a good idea to add `JavaType` objects to the recipe's accumulator in the first place. You should instead get the type's fully qualified name using `JavaType.FullyQualified#getFullyQualifiedName()` and add these strings to the set.

Reply:
done